Implement asset page

Gudahtt opened the pull request:
A new page has been created for viewing assets. This replaces the old `selectedToken` state, which previously would augment the home page to show token-specific information.

The new asset page shows the standard token overview as seen previously on the home page, plus a history filtered to show just transactions relevant to that token.

The actions that were available in the old token list menu have been moved to a "Token Options" menu that mirrors the "Account Options" menu.

The `selectedTokenAddress` state has been removed, as it is no longer being used for anything.

A new Redux store has been added to track state related to browser history. The most recent "overview" page (i.e. the home page or the asset page) is currently being tracked, so that actions taken from the asset page can return the user back to the asset page when the action has finished.

Gudahtt committed:
Implement asset page

A new page has been created for viewing assets. This replaces the old
`selectedToken` state, which previously would augment the home page
to show token-specific information.

The new asset page shows the standard token overview as seen previously
on the home page, plus a history filtered to show just transactions
relevant to that token.

The actions that were available in the old token list menu have been
moved to a "Token Options" menu that mirrors the "Account Options"
menu.

The `selectedTokenAddress` state has been removed, as it is no longer
being used for anything.

`getMetaMetricState` has been renamed to `getBackgroundMetaMetricState`
because its sole purpose is extracting data from the background state
to send metrics from the background. It's not really a selector, but
it was convenient for it to use the same selectors the UI uses to
extract background data, so I left it there for now.

Gudahtt commented:
This depends upon #8695

Gudahtt committed:
Add history store

The `history` store tracks state related to browser history. The most
recent "overview" page (i.e. the home page or the asset page) is
currently being tracked, so that actions taken from the asset page can
return the user back to the asset page when the action has finished.

Gudahtt committed:
Create separate action for changing send token

This action also handles clearing other `send` state relating to tokens
when switching back to the native currency.

Gudahtt commented:
<details>
<summary>Screenshots:</summary>
Home (popup):

![home-popup](https://user-images.githubusercontent.com/2459287/83314468-f140cc00-a1f0-11ea-8365-5ee58235be8a.png)

Asset page - Eth (popup):

![eth-popup](https://user-images.githubusercontent.com/2459287/83314482-0289d880-a1f1-11ea-9cc4-aa2ec450c0e2.png)

Asset page - TST token (popup):

![tst-popup](https://user-images.githubusercontent.com/2459287/83314491-0ddd0400-a1f1-11ea-8cda-a2f5581dc4b1.png)

Asset page - TST token with menu open (popup):

![tst-menu-popup](https://user-images.githubusercontent.com/2459287/83314499-1cc3b680-a1f1-11ea-9e78-0036dd295152.png)

Home (fullscreen):

![home-fullscreen](https://user-images.githubusercontent.com/2459287/83314508-24835b00-a1f1-11ea-8099-d99567013ce0.png)

Asset page - Eth (fullscreen):

![eth-fullscreen](https://user-images.githubusercontent.com/2459287/83314521-3107b380-a1f1-11ea-97d3-13a6992aed68.png)

Asset page - TST token with menu open (fullscreen):

![test-menu-fullscreen](https://user-images.githubusercontent.com/2459287/83314530-3a911b80-a1f1-11ea-9b29-ab340f851ad9.png)

</details>

rekmarks commented:
Fantastic!

Were there any designs for the asset rows being highlighted on hover, with a link/pointer cursor? I imagine we want at least the pointer cursor, but a subtle highlight would be best, IMO.

Gudahtt commented:
Yes, good point! The hover state was in the design and is missing. Cursor pointer seems appropriate as well. I'll add those

Gudahtt committed:
Update asset list item styles

The asset list items now have borders, a grey background on hover, and
a pointer cursor.

Gudahtt commented:
I've updated the asset list items to have a grey background on hover, `cursor: pointer`, and a border similar to the transaction list items.

These things are still TODO:
* Add the inline "Send [token]" button on the asset list items, but only on large viewports
* Use the `ListItem` class for both lists (some minor refactoring is required first here), and update the styles to get more consistent heights, spacing, etc. across both lists
* Update tab styles to match design more closely
* Expand the "warning" state to show the entire warning inline on large viewports

None of these seem trivial, nor do they seem like blockers, so I'd prefer to deal with them in a future PR if possible to avoid blocking the RC, and to keep this monstrously sized PR from getting any larger

rekmarks commented on ui/app/ducks/metamask/metamask.js:
```diff
@@ -227,6 +199,35 @@ export default function reduceMetamask (state = {}, action) {
+          },
+        })
+      }
+      return Object.assign(metamaskState, {
```
nit: Not sure whether we have a preference between spreading and `Object.assign`, but if we do, we should probably follow one or the other?

Gudahtt replied:
Using the spread operator here wouldn't be functionally equivalent; this mutates `metamaskState` rather than constructing a new object. Which is fine here, since `metamaskState` is initialized to a new object at the start of the reducer. The equivalent of `{ ...x, ...y }` would be `Obejct.assign({}, x, y)` (with the empty object as the first parameter).

Ultimately I would prefer our convention to be neither of them, but to use `immer` instead. In the meantime either seems fine, with this being slightly better because it allocates one less unnecessary object 🤷‍♂️

rekmarks commented on ui/app/pages/asset/asset.js:
```diff
@@ -0,0 +1,65 @@
+  } else if (asset === nativeCurrency) {
+    assetName = nativeCurrency
+  } else {
+    return <Redirect to={{ pathname: DEFAULT_ROUTE }} />
```
This is if the user is directed to this page without an asset selected, right? Under what circumstances might that happen? The user navigating through the browser history?

Gudahtt replied:
Sure, it could happen if the user navigated there directly. More often it would be after switching accounts, switching networks,, or when a token was removed,

brad-decker commented:
Q: Should clicking on ETH show only ETH transactions? I know in the previous version of the UI clicking ETH was just like 'unselecting' a token, but with the new UI and the header "Account 1 / ETH" I'm not sure I'd expect token transactions to show up. -- Not at all a blocker for merge just curious.

Gudahtt commented:
>Q: Should clicking on ETH show only ETH transactions?

Yes, that was the intention for now, since all transactions use Eth for gas. It seemed strange to allow the eth balance to change without showing why it changed in the transaction history.

Maybe we should filter out signature requests in that view though? 🤔 🤷‍♂️ I hadn't thought of that, since they're only shown when pending.

brad-decker commented:
Perhaps. Maybe, in a later iteration, we can highlight token transfers on the eth screen in some way that makes it evident that they are token transfers, but that indicates the gas prices paid in eth.

Gudahtt commented:
Yeah, perhaps by showing the gas cost instead of the token balance that was transferred. I was just thinking that too. Not sure how to make clear that it's the gas price we're showing though, and not the  transfer amount converted to Eth.

brad-decker reviewed (dismissed):
LGTM

rekmarks commented on ui/app/pages/asset/asset.scss:
```diff
@@ -0,0 +1,50 @@
+    overflow-y: auto;
+    background-color: white;
+
+    &::-webkit-scrollbar {
```
We should probably add `scrollbar-width: none;` here as well? Currently the page has a scrollbar in Firefox.

Gudahtt replied:
Huh, that rule shouldn't be there at all. There shouldn't be an scrollbar here because the whole page scrolls instead. I'll remove this.

Gudahtt replied:
This has been addressed in d07d54b3516977b53a9dab949d0dbf6fcbb9820b

rekmarks replied:
Ah, so we want a scrollbar on the whole page?

rekmarks replied:
This was Firefox before your most recent commit:
![image](https://user-images.githubusercontent.com/25517051/83430236-249a7b00-a3ea-11ea-9809-fbc508e7fd51.png)

Gudahtt replied:
I believe this is a pre-existing discrepancy between Chrome and Firefox. That scrollbar is not there due to the `asset__container` element as far as I can tell.

Gudahtt replied:
I was able to reproduce this same scrollbar on the home page as well by adding many assets.

I don't entirely understand why the scrollbar is hidden in Chrome, but not Firefox 🤔 There is `webkit-scrollbar` rule on `main-container` to disable the scrollbar on Chrome, but adding `scrollbar-width: none` to the same class seemed to have no effect. I had to add `scrollbar-width: none` to `.app` to disable it for some reason.

Gudahtt committed:
Remove unnecessary CSS rules

The `.asset__container` element doesn't need to set `overflow-y` or any
scrollbar-related rules because that element doesn't have a defined
`height` or `max-height`; it can grow as long as it wants. There can
be no overflow, and no scrollbar.

brad-decker reviewed (approved):
LGTM

rekmarks commented:
Why do we use the history state/ducks instead of `history.goBack()`?

Gudahtt commented:
`history.goBack` doesn't work because we use the `HashRouter`, which doesn't keep any history.

rekmarks reviewed (approved):
Great work!

brad-decker commented:
:shipit:
